rcl_yaml_param_parser/src/parser.c

```diff
@@ -0,0 +1,1165 @@
+// Copyright 2018 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <yaml.h>
+
+#include "rcl_yaml_param_parser/parser.h"
+#include "rcl_yaml_param_parser/types.h"
+
+#include "rcl/error_handling.h"
+#include "rcl/types.h"
+#include "rcutils/strdup.h"
+
+/// NOTE: Will allow a max YAML mapping depth of 5
+/// map level 1 : Top level mapping of the config file
+/// map level 2 : Node namespace mapping
+/// map level 3 : Node name mapping
+/// map level 4 : Params mapping
+/// map level 5 : Params namespace mapping
+typedef enum yaml_map_lvl_e
+{
+  MAP_UNINIT_LVL = 0U,
+  MAP_TOP_LVL = 1U,
+  MAP_NODE_NS_LVL = 2U,
+  MAP_NODE_NAME_LVL = 3U,
+  MAP_PARAMS_LVL = 4U,
+  MAP_PARAMS_NS_LVL = 5U
+} yaml_map_lvl_t;
```

Comment:
Is this implementation robust to multiple namespacing levels? e.g.:

```
stereo_cameras_ns_top:
  stereo_cameras_ns_low:
    stereo_camera1:
      params:
        is_cam_on: [true, true, false, true, false, false]
        brand: Bosch
    stereo_camera2:
      params:
        stereo_camera_dr:
          stereo_camera_left:
            image_type: ir
          stereo_camera_right:
            image_type: rbg
```

Reply:
Oh I must have messed up something then. I tried just adding one level of namespacing to the correct_config.yaml file and it fails to parse on my machine. Do you have an idea of what I did wrong ?

```
diff --git a/rcl_yaml_param_parser/test/correct_config.yaml b/rcl_yaml_param_parser/test/correct_config.yaml
index 4b5e50d..51954f8 100644
--- a/rcl_yaml_param_parser/test/correct_config.yaml
+++ b/rcl_yaml_param_parser/test/correct_config.yaml
@@ -38,6 +38,7 @@ new_camera_ns:
       brand: Bosch
   new_camera2:
     params:
-      camera_dr:
-        dr_name: default
-      brand: Mobius 
+      camera_left:
+        camera_dr:
+          dr_name: default
+        brand: Mobius 
```

The test fail with the following message:
```
1: Will not support deeper mappings at line 43, at /home/mikael/work/ros2/bouncy_ws/src/ros2/rcl/rcl_yaml_param_parser/src/parser.c:1079
1: /home/mikael/work/ros2/bouncy_ws/src/ros2/rcl/rcl_yaml_param_parser/test/test_parse_yaml.cpp:37: Failure
1: Value of: res
1:   Actual: false
1: Expected: true

```

Reply:
Sorry my earlier answer is not entirely true. I restrict it to one namespacing level after params but can have mutilpe namespaces under params.  Things can get crazy allowing multiple namespace levels. Looked at quite a few examples in the config files in various places like autoware stacks, our internal lidar stack, and did not find a need for many namespace levels. Trying to keep simple

Reply:
It is very common in ROS to have several levels of namespacing for parameters.
For example, a minimalistic pan-tilt camera controller:
```
gimbal_controller:
  pan_joint:
    pid: {p: 50.0, i: 0, d: 2}
  tilt_joint:
    pid: {p: 50.0, i: 0, d: 2}
```

Reply:
Yes, it supports just one mapping level(params namespace mapping) after params. In the above example it has two mapping levels after params `camera_left` and `camera_dr`. It can have 

```
   new_camera2:
     params:
      camera_left_dr:
        dr_name: default
        brand: Mobius 
      camera_right_dr:
        dr_name: default
        brand: Mobius 
```

Reply:
We in the design decided with the following format. A parameter_namespace_string and series of `<field_name> : <value>` pairs. 
```
<node_namespace_string>:  # optional
  <node1_name>:
    params:
      <field_name>: <field_value>
      <parameter_namespace_string>:   # optional
        <field1_name>: <field1_value>
        <field2_name>: <field2_value>
  <node2_name>:
    params:
      <field_name>: <field_value>
      <parameter_namespace_string>:   # optional
        <field1_name>: <field1_value>
        <field2_name>: <field2_value>
```
So the parameter name we decided was `parameter_namespace_string.field_name`.  If we want to change this behavior, we can put it as a TODO and add more support after the basic support is done.

Just to keep in mind, a libyaml C parser library is a very very minimalistic parser. It is more of a tokeniser. Just parses each word and returns it. It does not provide any more info and the rcl parser code has to keep track of everything else a yaml specification talks about. It does not even tell if a returned token is a key or a value.  Doing multiple levels of mappings, can make the rcl_yaml_parser code very cumbersome. That is why there are lot of other wrappers written around libyamlc to parse fancy things. The biggest advantage of using libyaml C is, it will be comparatively easier to get it through MISRA compared to other parsers. It still has around 3000 MISRA complaints. 

Again, if you think we should have mutiple namespace levels and mutiple layers of `key:value` pairs inside a value like  `pid: {p: 50.0, i: 0, d: 2}`, we can put as a TODO and support more adavanced features later on

Reply:
Mikael, where can I find the code where mutiple level of parameter namespace gets used  (Eg: the gimbal_controller) . Just for me to get an understanding. Do you have a pointer to the code? Thanks